var/spack/repos/builtin/packages/cnpy/package.py

```diff
@@ -16,6 +16,8 @@ class Cnpy(CMakePackage):
 
     version('master', branch='master')
 
+    depends_on('zlib', type='link')
```

Comment:
```suggestion
    depends_on('zlib')
```
They include `zlib.h` in their headers, so the default setting should be fine.

Reply:
But set_build_environment_variable() in lib/spack/spack/build_environmeny.py is added headers directories of link type package to SPACK_INCLUDE_DIRS environment variable.

Reply:
@tkameyama I get that the package builds nonetheless, but I think it's more appropriate to use the default `('build', 'link')` dependency type as `zlib` is needed at build and link time. I see that the code you refer to has been modified last by @scheibelp so wondering if he's of a different opinion.

The reason I insist on this is e.g. [this header](https://github.com/rogersce/cnpy/blob/4e8810b1a8637695171ed346ce68f6984e585ef4/cnpy.h#L15-L17).

Reply:
Personally I think `type='link'` is better, as it is more specific. I'm also fine with the default settings, but I think `type='build'` is unused here.

Reply:
@adamjstewart What's the meaning of `build` then? `zlib.h` is used when compiling translation units so I think that's a build dep too. Or the meaning has shifted at some point to only tooling? What about header only libraries?

Reply:
My understanding is that `build` means "add this package to `PATH` and `PYTHONPATH`". `link` does double duty and serves as both `-L` and `-I` flags for the compiler wrappers.

Reply:
I raised this issue with @scheibelp once. I feel like I remember him creating a PR to add an "include" deptype that handled `-I` but could be uninstalled after installation like "build" deptypes, but that PR never went anywhere.

Reply:
> My understanding is that build means "add this package to PATH and PYTHONPATH". link does double duty and serves as both -L and -I flags for the compiler wrappers.

That is what it does, not necessarily what it means :smirk: In any case if you're right most of the library dependencies are modeled wrong.

Reply:
For now, depending on `.h` files should impose a `link` deptype. The details are at https://github.com/spack/spack/issues/3768#issuecomment-293046651 but in short, depending on `.h` files imposes consistency constraints that only the `link` deptype respects right now (although it is heavy-handed).

The `build` deptype is for packages providing executables that are run at build time (like `CMake`).

Reply:
@adamjstewart It seems the `-I` behavior of `link` is not documented, but all the other things are like you said. Reference [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types). I'm a bit surprised that in our model C++ header only libraries are considered `link` dependencies, as (external) linking is what you want to avoid. I'll remove my red flag - merge at your will.

Reply:
@scheibelp Apologies, I didn't see your comment before refreshing the window. On a second read it seems this is not what is documented, so perhaps we need to update the documentation. Do I understand correctly then that `build` deps have lost the meaning of "needed at compile time" and became at some point just "tooling"?

Reply:
> Do I understand correctly then that build deps have lost the meaning of "needed at compile time" and became at some point just "tooling"?

Yes (more on that reasoning at https://github.com/spack/spack/pull/13016#issuecomment-537611179 and #3768). Let me know if you think that causes problems.